internal/api/core/middleware/proxy.go

```diff
@@ -1,118 +1,225 @@
+			return nil, "", echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("unable to forward the request to the datasource '%s', datasource doesn't exist", datasourceName))
+		}
+		logrus.WithError(err).Errorf("unable to find the datasource '%s', something wrong with the database", datasourceName)
+		return nil, "", echo.NewHTTPError(http.StatusInternalServerError, "internal server error")
+	}
+	// Based on the HTTP 1.1 RFC, a `/` should be the minimum path.
+	// https://datatracker.ietf.org/doc/html/rfc2616#section-5.1.2
+	path := "/"
+	if len(matchingGroups[0]) > 1 {
+		path = matchingGroups[0][2]
+	}
+	return dts.Spec, path, nil
+}
+
+type proxy interface {
+	serve(c echo.Context) error
+}
+
+func newProxy(spec v1.DatasourceSpec, path string) (proxy, error) {
+	switch v := spec.(type) {
+	case *datasourcev1.Prometheus:
+		return &httpProxy{config: v.HTTP, path: path}, nil
+	default:
+		return nil, echo.NewHTTPError(http.StatusBadGateway, fmt.Sprintf("datasource type '%T' not managed", spec))
+	}
 }
 
-func serveProxy(c echo.Context, targetURL *url.URL) error {
+type httpProxy struct {
+	config datasourcev1.HTTPConfiguration
+	path   string
+}
+
+func (h *httpProxy) serve(c echo.Context) error {
 	req := c.Request()
 	res := c.Response()
 
+	if err := h.prepareRequest(c); err != nil {
+		return err
+	}
+
+	// redirect the request to the datasource
+	req.URL.Path = h.path
+
+	// Set up the proxy
+	var proxyErr error
+	reverseProxy := httputil.NewSingleHostReverseProxy(h.config.URL)
+	reverseProxy.ErrorHandler = func(writer http.ResponseWriter, request *http.Request, err error) {
+		desc := h.config.URL.String()
+		proxyErr = fmt.Errorf("error proxying, remote unreachable: target=%s, err=%w", desc, err)
+		logrus.Errorf(proxyErr.Error())
+		proxyErr = err
+	}
+	// use a dedicated HTTP transport to avoid any TSL encrypt issue
+	reverseProxy.Transport = h.prepareTransport()
+	// Reverse proxy request.
+	reverseProxy.ServeHTTP(res, req)
+	// Return any error handled during proxying request.
+	return proxyErr
+}
+
+func (h *httpProxy) prepareRequest(c echo.Context) error {
+	req := c.Request()
 	// We have to modify the HOST of the request in order to match the host of the targetURL
 	// So far I'm not sure to understand exactly why, but if you are going to remove it, be sure of what you are doing.
 	// It has been done to fix an error returned by Openshift itself saying the target doesn't exist.
 	// Since we are using HTTP/1, setting the HOST is setting also an header so if the host and the header are different
 	// then maybe it is blocked by the Openshift router.
-	req.Host = targetURL.Host
+	req.Host = h.config.URL.Host
 	// Fix header
 	if len(req.Header.Get(echo.HeaderXRealIP)) == 0 {
 		req.Header.Set(echo.HeaderXRealIP, c.RealIP())
 	}
 	if len(req.Header.Get(echo.HeaderXForwardedProto)) == 0 {
 		req.Header.Set(echo.HeaderXForwardedProto, c.Scheme())
 	}
+	// set header according to the configuration
+	if len(h.config.Headers) > 0 {
```

Comment:
Do we want to allow overriding even special headers like the user agent or the host header here? Maybe fine for now, but I'd add a TODO to think about this later.

Reply:
indeed we shouldn't have a list of header that cannot be overridden

Reply:
will put a todo